src/libopensc/card-idprime.c

```diff
@@ -0,0 +1,789 @@
+#include "pkcs15.h"
+
+static const struct sc_card_operations *iso_ops = NULL;
+
+static struct sc_card_operations idprime_ops;
+static struct sc_card_driver idprime_drv = {
+	"Gemalto IDPrime",
+	"idprime",
+	&idprime_ops,
+	NULL, 0, NULL
+};
+
+/* This ATR says, there is no EF.DIR nor EF.ATR so ISO discovery mechanisms
+ * are not useful here */
+static const struct sc_atr_table idprime_atrs[] = {
+	{ "3b:7f:96:00:00:80:31:80:65:b0:84:41:3d:f6:12:0f:fe:82:90:00",
+	  "ff:ff:00:ff:ff:ff:ff:ff:ff:ff:00:00:00:00:ff:ff:ff:ff:ff:ff",
+	  "Gemalto IDPrime MD 8840, 3840, 3810, 840 and 830 Cards",
+	  SC_CARD_TYPE_IDPRIME_GENERIC, 0, NULL },
+};
+
+static const sc_path_t idprime_path = {
+	"", 0,
+	0, 0, SC_PATH_TYPE_DF_NAME,
+	{ "\xA0\x00\x00\x00\x18\x80\x00\x00\x00\x06\x62", 11 }
+};
+
+/* data structures to store meta data about IDPrime objects */
+typedef struct idprime_object {
+	int fd;
+	unsigned char key_reference;
+	u8 df[2];
+	unsigned short length;
+} idprime_object_t;
+
+/*
+ * IDPrime private data per card state
+ */
+typedef struct idprime_private_data {
+	u8 *cache_buf;			/* cached version of the currently selected file */
+	size_t cache_buf_len;		/* length of the cached selected file */
+	int cached;			/* is the cached selected file valid */
+	size_t file_size;		/* this is real file size since IDPrime is quite strict about lengths */
+	list_t pki_list;		/* list of pki containers */
+	idprime_object_t *pki_current;	/* current pki object _ctl function */
+	u8 tinfo_df[2];			/* DF of object with Token Info Label */
+} idprime_private_data_t;
+
+/* For SimCList autocopy, we need to know the size of the data elements */
+static size_t idprime_list_meter(const void *el) {
+	return sizeof(idprime_object_t);
+}
+
+void idprime_free_private_data(idprime_private_data_t *priv)
+{
+	free(priv->cache_buf);
+	list_destroy(&priv->pki_list);
+	free(priv);
+	return;
+}
+
+idprime_private_data_t *idprime_new_private_data(void)
+{
+	idprime_private_data_t *priv;
+
+	priv = calloc(1, sizeof(idprime_private_data_t));
+	if (priv == NULL)
+		return NULL;
+
+	/* Initialize PKI Applets list */
+	if (list_init(&priv->pki_list) != 0 ||
+	    list_attributes_copy(&priv->pki_list, idprime_list_meter, 1) != 0) {
+		idprime_free_private_data(priv);
+		return NULL;
+	}
+
+	return priv;
+}
+
+int idprime_add_object_to_list(list_t *list, const idprime_object_t *object)
+{
+	if (list_append(list, object) < 0)
+		return SC_ERROR_INTERNAL;
+	return SC_SUCCESS;
+}
+
+/* This selects main IDPrime AID which is used for communication with
+ * the card */
+static int idprime_select_idprime(sc_card_t *card)
+{
+	return iso_ops->select_file(card, &idprime_path, NULL);
+}
+
+/* This select some index file, which is useful for enumerating other files
+ * on the card */
+static int idprime_select_index(sc_card_t *card)
+{
+	int r;
+	sc_file_t *file = NULL;
+	sc_path_t index_path;
+
+	/* First, we need to make sure the IDPrime AID is selected */
+	r = idprime_select_idprime(card);
```

Comment:
This is the first select in the card detection (and then in the initialization). In the first case, we know it is not selected. In the second run, we already selected the index.

I think I wrote this really as a sanity check to make sure we talk to the correct card, but having a second look now, it might be better to leave this only in the card detection step. But I would have to check with real card.